### DISCUSSION TITLE
Can I use PSRAM as additional memory on homespan?

### DISCUSSION THREAD

**Discussion by Nezaemmy** [HomeSpan/HomeSpan#969] [Q&A]
I created code to run 20 outputs with buttons to control these outputs using single board with shift registers.
I'm using Esp32s3 with 8M PSRAM. Is homespan can access on the PSRAM to use it as additional to internal RAM of esp32s3?

**Comment by HomeSpan** [comment]
Yes, but you need to enable PSRAM use from within the Arduino Interface for your chip.  HomeSpan automatically uses whatever is available. Note depending on the chip there MAY be different settings for how PSRAM is accessed when enabled - you will see those choices in the Arduino Interface.  Depending on the choice, the chip may use up some additional pins that you won't be able to use for other features.  You'll have to consult any documentation for your specific chip.

**Reply by Nezaemmy** [reply]
Thank you

**Comment by Nezaemmy** [comment]
> Yes, but you need to enable PSRAM use from within the Arduino Interface for your chip.  HomeSpan automatically uses whatever is available. Note depending on the chip there MAY be different settings for how PSRAM is accessed when enabled - you will see those choices in the Arduino Interface.  Depending on the choice, the chip may use up some additional pins that you won't be able to use for other features.  You'll have to consult any documentation for your specific chip. 

@HomeSpan did you use PSRAM? I tried to use PSRAM on my project but I can't find the way to use it on my homespan code.
Do you have any suggestion of how to do it?

**Reply by HomeSpan** [reply]
Yes, I use PSRAM all the time.  I don't understand your question above.

There is nothing *in HomeSpan* you need to do to use PSRAM.  BUT...you still need to (a) have a chip with PSRAM (not all do), and (b) enable PSRAM when you compile your code.   Again, enabling PSRAM has nothing to do with HomeSpan.  You enable it the same way you enable it for any sketch of any type. In the Arduino IDE there is a Tools menu where you select your Port , Board, Partition, etc.  It's right there in the menu --- it says PSRAM: Enable/Disable.  Select Enable to enable.  **If you don't see that menu selection it means you don't have any PSRAM on your chip.**

If you you are using Platform I/O instead of the Arduino IDE, you'll have to ask others.  I don't use that platform and don't know where the menu selection would be for enabling PSRAM.

**Reply by Nezaemmy** [reply]
Thank you @HomeSpan for your help.
I use esp32s3 with 8M PSRAM, and PSRAM enable in Arduino IDE-tools. but it seems like to enable in tools is not enough to allocate the memory  https://github.com/espressif/arduino-esp32/discussions/10782#discussioncomment-11671385 
I want to use PSRAM when memory buffer on my homespan project in not enaugh on DSRAM os esp32 because I use many outputs and buttons

**Reply by HomeSpan** [reply]
In the HomeSpan Serial Monitor type 'm'. What does it show?

**Reply by Nezaemmy** [reply]
```
 Allocated      Free   Largest       Low
            --------- --------- --------- ---------
Total Heap:    118568   8590620   8257524   8586728
  Internal:     84176    244892    180212    243648
     PSRAM:     34392   8345728   8257524   8343080

Lowest stack level: 5308 bytes (loopTask)
NVS Flash Partition: 108 of 504 records used
```

**Reply by Nezaemmy** [reply]
> ```
> ```
> ```
> 
> Allocated Free Largest Low --------- --------- --------- --------- Total Heap: 118568 8590620 8257524 8586728 Internal: 84176 244892 180212 243648 PSRAM: 34392 8345728 8257524 8343080
> 
> Lowest stack level: 5308 bytes (loopTask) NVS Flash Partition: 108 of 504 records used

@HomeSpan I don't know if you can see it very well

**Comment by HomeSpan** [comment]
Okay, so this shows that HomeSpan will be using PSRAM for most (though not all) functions.  It may be that you are using up too much of the internal RAM and there are still things I can optimize in HomeSpan, but that's not very likely since you still have a lot of internal RAM free.

HOWEVER, that's only for the HomeSpan library.  If you are accessing other libraries in your sketch or creating any objects or even using "new" to create a Non-HomeSpan object, those Non-HomeSpan functions will use internal memory.  I'm afraid this is outside the scope of HomeSpan support - you will need to do some research about how to use PSRAM more generally in the ESP32 environment.  I think the link you cited above provides some good resources.

**Reply by Nezaemmy** [reply]
> Okay, so this shows that HomeSpan will be using PSRAM for most (though not all) functions. It may be that you are using up too much of the internal RAM and there are still things I can optimize in HomeSpan, but that's not very likely since you still have a lot of internal RAM free.
> 
> HOWEVER, that's only for the HomeSpan library. If you are accessing other libraries in your sketch or creating any objects or even using "new" to create a Non-HomeSpan object, those Non-HomeSpan functions will use internal memory. I'm afraid this is outside the scope of HomeSpan support - you will need to do some research about how to use PSRAM more generally in the ESP32 environment. I think the link you cited above provides some good resources.

Thank you so much for your support

**Reply by Nezaemmy** [reply]
> Okay, so this shows that HomeSpan will be using PSRAM for most (though not all) functions. It may be that you are using up too much of the internal RAM and there are still things I can optimize in HomeSpan, but that's not very likely since you still have a lot of internal RAM free.
> 
> HOWEVER, that's only for the HomeSpan library. If you are accessing other libraries in your sketch or creating any objects or even using "new" to create a Non-HomeSpan object, those Non-HomeSpan functions will use internal memory. I'm afraid this is outside the scope of HomeSpan support - you will need to do some research about how to use PSRAM more generally in the ESP32 environment. I think the link you cited above provides some good resources.

I created big project use 13 buttons and 6 motion sensors, and 20 outputs with only one board esp32s3.
before I used spanButton for those buttons and sensors, the memory was total not enough, and the system was hanging always.
Now I replaced spanButton with 3rd party library to control buttons and sensors; now there is improvement of the memory the system use.
